### PR TITLE
Changelog

### PR DESCRIPTION
**Pull request by santokhsingh**
- JSGrid integration on admin pages
- Server side pagination and filtering on applications and users page
- Application description field

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -64,6 +64,7 @@ app.use(logger('{"date":":date[clf]","method":":method","url":":url","remote-add
 app.use(express.static(path.join(__dirname, 'public')));
 app.use('/assets/bootstrap', express.static(path.join(__dirname, 'node_modules/bootstrap/dist')));
 app.use('/assets/jquery', express.static(path.join(__dirname, 'node_modules/jquery/dist')));
+app.use('/assets/jsgrid', express.static(path.join(__dirname, 'node_modules/jsgrid/dist')));
```
Comment (DonMartin76) [inline app.js]: 👍

File: routes/admin.js [hunk 2]
```diff
@@ -94,7 +94,10 @@ router.get('/users', function (req, res, next) {
     //   filter: URI encoded JSON, e.g. {"name":"michael"}
     //   order_by: e.g. name%20DESC, or customId%20ASC
     //   no_cache: 0 or 1, set to 1 to really count the records, otherwise the result may be read from a cache
-    utils.getFromAsync(req, res, '/registrations/pools/wicked', 200, function (err, apiResponse) {
+    var filterFields  = ['id','name','email'];
```
Comment (DonMartin76) [inline routes/admin.js]: Please use `const` or `let` for new code; I am phasing out all usage of `var`. Thanks.

File: routes/admin.js [hunk 3]
```diff
@@ -94,7 +94,10 @@ router.get('/users', function (req, res, next) {
...
+    var usersUri = utils.makePagingUri(req, '/registrations/pools/wicked?embed=1',filterFields);
```
Comment (DonMartin76) [inline routes/admin.js]: `?embed=1` will not do anything here; only for the `/applications` end point, where it just returns a list of IDs if you omit it.

File: routes/admin.js [hunk 4]
```diff
@@ -139,59 +136,29 @@ router.get('/applications', function (req, res, next) {
     //       sorting cannot be done sensibly.
     //
     // Typical request:
-    // 
+    //
     //       GET /applications?embed=1&filter=...&order_by=name%20ASC&offset=0&limit=20
-    utils.getFromAsync(req, res, '/applications', 200, function (err, appsResponse) {
+    var filterFields  = ['id','name','ownerEmail'];
+    var appsUri = utils.makePagingUri(req, '/applications?embed=1',filterFields);
```
Comment (DonMartin76) [inline routes/admin.js]: Please always run a code formatter over your code; if using VS Code, it's just Alt+Shift+F (here: Missing space after comma...). <nitpick>

File: routes/admin.js [hunk 5]
```diff
@@ -139,59 +136,29 @@ router.get('/applications', function (req, res, next) {
...
-            appIds.push(appsResponse.items[i].id);
-
-        // This is the expensive part:
-        async.map(appIds, function (appId, callback) {
-            utils.getFromAsync(req, res, '/applications/' + appId, 200, callback);
-        }, function (err, appsInfos) {
-            if (err)
-                return next(err);
-            for (let i = 0; i < appsInfos.length; ++i) {
-                const thisApp = appsInfos[i];
```
Comment (DonMartin76) [inline routes/admin.js]: Yay for throwing out code!

File: routes/utils.js [hunk 6]
```diff
@@ -57,6 +57,25 @@ utils.ensureNoSlash = function (url) {
     return url;
 };
 
+utils.makePagingUri = function (req, uri, filterFields) {
+    var startIndex = (req.query.pageIndex && req.query.pageSize) ?  (req.query.pageIndex - 1) * req.query.pageSize : 0;
+    uri  = uri+ "&offset="+startIndex;
+    uri  = ( req.query.pageSize )  ? uri + "&limit="+req.query.pageSize : uri;
+    uri  = ( req.query.sortField ) ? uri +"&order_by="+req.query.sortField+"%20"+req.query.sortOrder.toUpperCase() : uri;
```
Comment (DonMartin76) [inline routes/utils.js]: Rather use

const qs = require('querystring'); // might already be present here

... and the `qs.escape(whatever)` method for URI encoding. That's what is used all over the code.

I can also recommend using template strings for string formatting: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/template_strings
That mostly makes it easier to read.

File: views/layout.jade [hunk 7]
```diff
@@ -19,6 +19,9 @@ html
     link(rel='stylesheet', href='/assets/bootstrap/css/bootstrap.min.css')
     link(rel='stylesheet', href='/content/wicked.css')
     link(rel="stylesheet", href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.3.0/styles/default.min.css")
```
Comment (DonMartin76) [inline views/layout.jade]: While you're at it, would you mind also replacing highlight.js with its npm counterpart and also serving it via the `/assets` end point?

File: views/application.jade [hunk 8]
```diff
@@ -72,6 +72,9 @@ block content
                             .form-group
                                 label(for='appname') Application Name:
                                 input(value=application.name name='appname').form-control
+                            .form-group
+                                label(for='appdesc') Application Description:
+                                input(value=application.description name='appdesc').form-control
```
Comment (DonMartin76) [inline views/application.jade]: I would rather have a div here with Markdown enabled content; `marked` should be present at least on the backend, so using that here should be really simple. Use `!{someVariable}` in the Jade code to not HTML encode the content, but take the things as is.

Is there a reason why we should not enable Jade here?

**Comment by santokhsingh** [comment]
Ok, yeah agree on the qs.escape it was just lazy on my part. Formatting i tried to do with the VScode, is it still an issue?

**Pull request merged**